### PR TITLE
Incremental backup and point in time recovery for XtraBackup 

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Followup to https://github.com/vitessio/vitess/pull/11097

This PR supports incremental backup and point in time recoveries for backups using `--backup_engine_implementation=xtrabackup`.

Incremental backup is still done by copying `mysqlbinlog` files (thus we do not use `XtraBackup`'s incremental backup method of copying InnoDB pages). Point in time recovery is likewise done by replaing binary log files.

In fact, incremental backup and PITR are done using the `BuiltIn` backup&restore engine. So, for example, if you're running a point in time recovery using `xtrabackup`, the full backup is restored via `xtrabackup` engine, but then the binlog files are replayed by `builtin` engine.

Best reviewed [with spaces ignored](https://github.com/vitessio/vitess/pull/13156/files?w=1).

To be documented.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/11227

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/test/endtoend/backup/pitr/backup_mysqlctld_pitr_test.go [hunk 1]
```diff
@@ -52,161 +52,181 @@ func waitForReplica(t *testing.T) {
 }
 
 // TestIncrementalBackupMysqlctld - tests incremental backups using myslctld
-func TestIncrementalBackupMysqlctld(t *testing.T) {
```
Comment (shlomi-noach) [inline go/test/endtoend/backup/pitr/backup_mysqlctld_pitr_test.go:55]: Most of the content here is the same, just indented, and executes for any one of the three backup types we support. Best reviewed [with spaces ignored](https://github.com/vitessio/vitess/pull/13156/files?w=1).

**Comment by shlomi-noach** [comment]
Hmmm need to install `xtrabackup` for the CI job

**Review by deepthi — APPROVED** [review]
Nice work!
> In fact, incremental backup and PITR are done using the BuiltIn backup&restore engine. 

This is clever but possibly confusing.. but it doesn't seem worth pulling that code out into some sort of utility/shared package.

And just to make doubly sure - an incremental backup copies binlogs _without_ shutting down MySQL, right? That's how I'm reading the code, but want to be sure.

**Comment by shlomi-noach** [comment]
> an incremental backup copies binlogs without shutting down MySQL, right? 

That's correct.

**Comment by shlomi-noach** [comment]
Update: there are legit test failure with `backup_pitr` on `5.7`. This isn't specific to `5.7`. The scenario that happens, though is where a backup is taken while the very first binary log is open. This is a pretty unique scenario never expected in prod, and cannot happen with `builtin` (because `builtin` takes down the MySQL server, thereby rotating binary logs).

It's unique in that the incremental restore needs to extract the full backup, and then apply events from the very first binlog. There's just a bit more math to do because the first binary log has empty `Previous GTIDs` entry.

**Comment by shlomi-noach** [comment]
Narrowed down: seems like in a `5.7` restore (`Xtrabackup 2.4`), the restore process sets `@@gtid_purged` to be the value of `@@gtid_purged` at the time of the backup.

But this is no the desired value. We want to set `@@gtid_purged` to be the value of `@@gtid_executed` at the time of backup.

Fixing this.

**Comment by shlomi-noach** [comment]
> test failure looks related

fixed

**Comment by shlomi-noach** [comment]
Ahh, actually now it does show something relevant:

```
    backup_test.go:862: 
        	Error Trace:	/home/runner/work/vitess/vitess/go/vt/wrangler/testlib/backup_test.go:862
        	Error:      	Received unexpected error:
        	            	unexpected query: RESET MASTER
        	            	Issuing RESET MASTER
        	            	Can't restore backup
        	Test:       	TestDisableActiveReparents
```

File: go/vt/mysqlctl/backup.go [hunk 2]
```diff
@@ -291,6 +299,43 @@ func ShouldRestore(ctx context.Context, params RestoreParams) (bool, error) {
 	return checkNoDB(ctx, params.Mysqld, params.DbName)
 }
 
+// ensureRestoredGTIDPurgedMatchesManifest sees o the following: when you restore a full backup, you want the MySQL server to have
```
Comment (mattlord) [inline go/vt/mysqlctl/backup.go:302]: I think we have an errant "o" here: `sees o the`

File: go/vt/mysqlctl/backup.go [hunk 3]
```diff
@@ -291,6 +299,43 @@ func ShouldRestore(ctx context.Context, params RestoreParams) (bool, error) {
...
+	gtid := manifest.Position.GTIDSet.String()
+	if gtid == "" {
+		return nil
+	}
+	// Xtrabackup 2.4's restore seems to set @@gtid_purged to be the @@gtid_purged at the time of backup. But this is not
+	// the desired value. We want to set @@gtid_purged to be the @@gtid_executed of the backup.
+	// As reminder, when restoring from a full backup, setting @@gtid_purged also sets @@gtid_executed.
+	restoredGTIDPurgedPos, err := params.Mysqld.GetGTIDPurged(ctx)
+	if err != nil {
+		return vterrors.Wrapf(err, "Reading gtid_purged after restore")
```
Comment (mattlord) [inline go/vt/mysqlctl/backup.go:321]: Nit, but error messages should not be capitalized.
```suggestion
		return vterrors.Wrapf(err, "failed to read gtid_purged after restore")
```

File: go/vt/mysqlctl/backup.go [hunk 4]
```diff
@@ -291,6 +299,43 @@ func ShouldRestore(ctx context.Context, params RestoreParams) (bool, error) {
...
+	}
+	if restoredGTIDPurgedPos.Equal(manifest.Position) {
+		return nil
+	}
+	params.Logger.Infof("Restore: @@gtid_purged does not equal manifest's GTID position. Setting @@gtid_purged to %v", gtid)
+	// This is not good. We want to apply a new @@gtid_purged value.
+	query := "RESET MASTER" // required dialect in 5.7
+	if _, err := params.Mysqld.FetchSuperQuery(ctx, query); err != nil {
+		return vterrors.Wrapf(err, "Issuing %v", query)
```
Comment (mattlord) [inline go/vt/mysqlctl/backup.go:330]:
```suggestion
		return vterrors.Wrapf(err, "error issuing %v", query)
```

File: go/vt/mysqlctl/backup.go [hunk 5]
```diff
@@ -291,6 +299,43 @@ func ShouldRestore(ctx context.Context, params RestoreParams) (bool, error) {
...
+	}
+	query = fmt.Sprintf("SET GLOBAL gtid_purged='%s'", gtid)
+	if _, err := params.Mysqld.FetchSuperQuery(ctx, query); err != nil {
+		return vterrors.Wrapf(err, "Applying`%s` after restore", query)
```
Comment (mattlord) [inline go/vt/mysqlctl/backup.go:334]:
```suggestion
		return vterrors.Wrapf(err, "failed to apply `%s` after restore", query)
```

File: go/vt/mysqlctl/binlogs_gtid.go [hunk 6]
```diff
@@ -108,6 +107,15 @@ func ChooseBinlogsForIncrementalBackup(
 		if err != nil {
 			return nil, "", "", vterrors.Wrapf(err, "cannot evaluate incremental backup from pos")
 		}
+		if incrementalBackupFromGTID == "" {
+			// This can happen on the very first binary log file. It happens in two scenarios:
+			// 1. This is the first binlog ever in the history of the mysql server; the GTID is truly empty
+			// 2. A full backup was taken and restored, with all binlog scraped.
```
Comment (mattlord) [inline go/vt/mysqlctl/binlogs_gtid.go:113]: Did you mean `binlog scrapped` here, meaning deleted? Or maybe you mean scraped as in read/applied?

File: go/vt/mysqlctl/mysqld.go [hunk 7]
```diff
@@ -1063,7 +1064,7 @@ socket=%v
 `, connParams.Uname, connParams.Pass, connParams.UnixSocket)
 	}
 
-	tmpfile, err := os.CreateTemp("", "example")
+	tmpfile, err := os.CreateTemp("", "defaults-extra-file-")
```
Comment (mattlord) [inline go/vt/mysqlctl/mysqld.go:1067]: Is the trailing dash here intentional?

File: go/vt/mysqlctl/mysqld.go [hunk 8]
```diff
@@ -1219,13 +1221,19 @@ func (mysqld *Mysqld) ApplyBinlogFile(ctx context.Context, binlogFile string, re
 		}
 		cnf, err := mysqld.defaultsExtraFile(params)
 		if err != nil {
-			return err
+			return vterrors.Wrapf(err, "creating default extra file")
```
Comment (mattlord) [inline go/vt/mysqlctl/mysqld.go:1224]:
```suggestion
			return vterrors.Wrapf(err, "failed to create defaults extra file")
```

File: go/vt/mysqlctl/mysqld.go [hunk 9]
```diff
@@ -1252,20 +1260,29 @@ func (mysqld *Mysqld) ApplyBinlogFile(ctx context.Context, binlogFile string, re
 		mysqlCmd.Dir = dir
 		mysqlCmd.Env = env
 		mysqlCmd.Stdin = pipe // piped from mysqlbinlog
+
+		mysqlCmd.Stderr = mysqlErrFile
+		log.Infof("ApplyBinlogFile: running mysql command: %#v with errfile=%v", mysqlCmd, mysqlErrFile.Name())
```
Comment (mattlord) [inline go/vt/mysqlctl/mysqld.go:1265]: Just confirming that we want go syntax `%#v` here vs printing out the struct with field names `%+v`. If so, then all good.

File: go/vt/mysqlctl/mysqld.go [hunk 10]
```diff
@@ -1252,20 +1260,29 @@ func (mysqld *Mysqld) ApplyBinlogFile(ctx context.Context, binlogFile string, re
...
 	}
 	// Run both processes, piped:
 	if err := mysqlbinlogCmd.Start(); err != nil {
 		return err
 	}
 	if err := mysqlCmd.Start(); err != nil {
-		return err
+		return vterrors.Wrapf(err, "starting mysql command")
```
Comment (mattlord) [inline go/vt/mysqlctl/mysqld.go:1272]:
```suggestion
		return vterrors.Wrapf(err, "failed to start mysql")
```

File: go/vt/mysqlctl/mysqld.go [hunk 11]
```diff
@@ -1252,20 +1260,29 @@ func (mysqld *Mysqld) ApplyBinlogFile(ctx context.Context, binlogFile string, re
...
 	}
 	// Wait for both to complete:
 	if err := mysqlbinlogCmd.Wait(); err != nil {
-		return err
+		return vterrors.Wrapf(err, "waiting on mysqlbinlog command")
```
Comment (mattlord) [inline go/vt/mysqlctl/mysqld.go:1276]:
```suggestion
		return vterrors.Wrapf(err, "mysqlbinlog command failed")
```

**Review by mattlord — APPROVED** [review]
LGTM! I only had a few minor questions/nits. I'll approve so that you can address them as you see fit.

File: go/vt/mysqlctl/mysqld.go (hunk 7 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/mysqld.go:1067]: yes - `os.CreateTemp()` then appends a random number to that.

File: go/vt/mysqlctl/mysqld.go (hunk 9 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/mysqld.go:1265]: Yeah, supposed to be `%#v` and this is just copied+pasted from similar calls elsewhere in this file. I'm just maintaining consistency.

**Comment by shlomi-noach** [comment]
All review comments were addressed.

**Pull request merged**